File: app/src/main/java/com/veganbeauty/app/features/profile/AccountProfilePersonalInfoFragment.java
package com.veganbeauty.app.features.profile;

import android.content.Context;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.veganbeauty.app.R;
import com.veganbeauty.app.core.base.RootieFragment;
import com.veganbeauty.app.data.local.ProfileSession;
import com.veganbeauty.app.databinding.AccountProfilePersonalInfoBinding;
import com.veganbeauty.app.utils.SyncDataHelper;

public class AccountProfilePersonalInfoFragment extends RootieFragment {

    private AccountProfilePersonalInfoBinding binding;
    private boolean isSaving;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = AccountProfilePersonalInfoBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void setupUI(@NonNull View view) {
        Context context = requireContext();
        loadFieldsFromSession(context);

        binding.etAddress.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {}

            @Override
            public void afterTextChanged(Editable s) {
                int length = s != null ? s.length() : 0;
                binding.tvAddressCount.setText(length + "/200");
            }
        });

        binding.btnBack.setOnClickListener(v -> {
            if (!isSaving) {
                getParentFragmentManager().popBackStack();
            }
        });

        binding.btnSave.setOnClickListener(v -> savePersonalInfo(false));

        com.veganbeauty.app.features.home.BottomNavHelper.highlightTab(view, R.id.nav_account);

        binding.layoutNotification.getRoot().setOnClickListener(v ->
                getParentFragmentManager().beginTransaction()
                        .replace(R.id.main_container, new com.veganbeauty.app.features.account.notification.AccountNotificationFragment())
                        .addToBackStack(null)
                        .commit());

        SyncDataHelper.syncUserProfileFromFirestore(context, () -> {
            if (binding != null && isAdded()) {
                loadFieldsFromSession(requireContext());
            }
        });
    }

    private void loadFieldsFromSession(Context context) {
        String fullName = ProfileSession.getFullName(context);
        String cccd = ProfileSession.getCCCD(context);
        String address = ProfileSession.getAddress(context);

        binding.etFullname.setText(fullName != null ? fullName : "");
        binding.etCccd.setText(cccd != null ? cccd : "");
        binding.etAddress.setText(address != null ? address : "");
        binding.tvAddressCount.setText((address != null ? address.length() : 0) + "/200");
    }

    private void savePersonalInfo(boolean exitAfterSave) {
        if (isSaving || binding == null) {
            return;
        }

        Context context = requireContext();
        String fullName = binding.etFullname.getText().toString().trim();
        String cccd = binding.etCccd.getText().toString().trim();
        String address = binding.etAddress.getText().toString().trim();

        if (fullName.isEmpty()) {
            Toast.makeText(context, "Họ và tên không được để trống", Toast.LENGTH_SHORT).show();
            binding.etFullname.requestFocus();
            return;
        }
        if (cccd.isEmpty()) {
            Toast.makeText(context, "Vui lòng nhập số CCCD", Toast.LENGTH_SHORT).show();
            binding.etCccd.requestFocus();
            return;
        }
        if (address.isEmpty()) {
            Toast.makeText(context, "Vui lòng nhập địa chỉ", Toast.LENGTH_SHORT).show();
            binding.etAddress.requestFocus();
            return;
        }

        ProfileSession.setFullName(context, fullName);
        ProfileSession.setCCCD(context, cccd);
        ProfileSession.setAddress(context, address);
        ProfileSession.markLocalProfileEdited(context);

        isSaving = true;
        binding.btnSave.setEnabled(false);
        Toast.makeText(context, "Đang lưu thông tin...", Toast.LENGTH_SHORT).show();

        SyncDataHelper.syncUserProfileToFirebaseAndLocal(context, (localSuccess, cloudSynced) -> {
            isSaving = false;
            if (binding == null || !isAdded()) {
                return;
            }
            binding.btnSave.setEnabled(true);
            if (!localSuccess) {
                Toast.makeText(context, "Không thể lưu thông tin. Vui lòng thử lại.", Toast.LENGTH_LONG).show();
                return;
            }
            if (!cloudSynced) {
                Toast.makeText(context, "Đã lưu trên máy. Đồng bộ cloud sẽ thử lại khi có mạng.", Toast.LENGTH_SHORT).show();
            }
            Toast.makeText(context, "Đã lưu thông tin cá nhân", Toast.LENGTH_SHORT).show();
            if (exitAfterSave) {
                getParentFragmentManager().popBackStack();
            }
        });
    }

    @Override
    protected void observeViewModel() {
        // Not used
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/com/veganbeauty/app/features/profile/AccountProfilePersonalInfoFragment.kt
package com.veganbeauty.app.features.profile

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.veganbeauty.app.core.base.RootieFragment
import com.veganbeauty.app.databinding.AccountProfilePersonalInfoBinding

class AccountProfilePersonalInfoFragment : RootieFragment() {

    private var _binding: AccountProfilePersonalInfoBinding? = null
    private val binding get() = _binding!!

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = AccountProfilePersonalInfoBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun setupUI(view: View) {
        val context = requireContext()
        val fullName = com.veganbeauty.app.data.local.ProfileSession.getFullName(context)
        val cccd = com.veganbeauty.app.data.local.ProfileSession.getCCCD(context)
        val address = com.veganbeauty.app.data.local.ProfileSession.getAddress(context)

        // Initial setup with masking
        binding.etFullname.setText(maskFullName(fullName))
        binding.etCccd.setText(maskCCCD(cccd))
        binding.etAddress.setText(maskAddress(address))
        binding.tvAddressCount.text = "${address.length}/200"

        // Set focus change listeners to handle secure editing
        binding.etFullname.setOnFocusChangeListener { _, hasFocus ->
            if (hasFocus) {
                val current = binding.etFullname.text.toString()
                if (current.contains("*")) {
                    binding.etFullname.setText(com.veganbeauty.app.data.local.ProfileSession.getFullName(context))
                }
            } else {
                val entered = binding.etFullname.text.toString()
                if (!entered.contains("*") && entered.isNotBlank()) {
                    com.veganbeauty.app.data.local.ProfileSession.setFullName(context, entered)
                }
                binding.etFullname.setText(maskFullName(com.veganbeauty.app.data.local.ProfileSession.getFullName(context)))
            }
        }

        binding.etCccd.setOnFocusChangeListener { _, hasFocus ->
            if (hasFocus) {
                val current = binding.etCccd.text.toString()
                if (current.contains("*")) {
                    binding.etCccd.setText(com.veganbeauty.app.data.local.ProfileSession.getCCCD(context))
                }
            } else {
                val entered = binding.etCccd.text.toString()
                if (!entered.contains("*") && entered.isNotBlank()) {
                    com.veganbeauty.app.data.local.ProfileSession.setCCCD(context, entered)
                }
                binding.etCccd.setText(maskCCCD(com.veganbeauty.app.data.local.ProfileSession.getCCCD(context)))
            }
        }

        binding.etAddress.addTextChangedListener(object : android.text.TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: android.text.Editable?) {
                val currentText = s?.toString() ?: ""
                if (binding.etAddress.hasFocus() && !currentText.contains("*")) {
                    binding.tvAddressCount.text = "${currentText.length}/200"
                } else {
                    val realAddr = com.veganbeauty.app.data.local.ProfileSession.getAddress(context)
                    binding.tvAddressCount.text = "${realAddr.length}/200"
                }
            }
        })

        binding.etAddress.setOnFocusChangeListener { _, hasFocus ->
            if (hasFocus) {
                val current = binding.etAddress.text.toString()
                if (current.contains("*")) {
                    binding.etAddress.setText(com.veganbeauty.app.data.local.ProfileSession.getAddress(context))
                }
            } else {
                val entered = binding.etAddress.text.toString()
                if (!entered.contains("*") && entered.isNotBlank()) {
                    com.veganbeauty.app.data.local.ProfileSession.setAddress(context, entered)
                }
                binding.etAddress.setText(maskAddress(com.veganbeauty.app.data.local.ProfileSession.getAddress(context)))
                binding.tvAddressCount.text = "${com.veganbeauty.app.data.local.ProfileSession.getAddress(context).length}/200"
            }
        }

        // Back button action
        binding.btnBack.setOnClickListener {
            binding.etFullname.clearFocus()
            binding.etCccd.clearFocus()
            binding.etAddress.clearFocus()
            parentFragmentManager.popBackStack()
        }

        // Highlight the "Tài khoản" tab as active in the bottom navigation menu
        view.findViewById<android.widget.LinearLayout>(com.veganbeauty.app.R.id.nav_account)?.let { navAccount ->
            val icon = navAccount.getChildAt(0) as? android.widget.ImageView
            val label = navAccount.getChildAt(1) as? android.widget.TextView
            
            // Set active green color tint to the icon (#677559)
            icon?.setColorFilter(android.graphics.Color.parseColor("#677559"))
            
            // Set active green color and bold style to the text label
            label?.setTextColor(android.graphics.Color.parseColor("#677559"))
            label?.setTypeface(null, android.graphics.Typeface.BOLD)
        }

        // Interactive Ux feedback Toasts
        binding.btnNotification.setOnClickListener {
            Toast.makeText(context, "Không có thông báo mới", Toast.LENGTH_SHORT).show()
        }
    }

    private fun maskFullName(fullName: String): String {
        if (fullName.isBlank()) return ""
        val firstChar = fullName.first().uppercase()
        val lastChar = fullName.last().uppercase()
        return "$firstChar*** **** ***$lastChar"
    }

    private fun maskCCCD(cccd: String): String {
        if (cccd.length < 4) return cccd
        return "*********${cccd.takeLast(4)}"
    }

    private fun maskAddress(address: String): String {
        if (address.isBlank()) return ""
        return "********"
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
